.github/workflows/format.yml

```diff
@@ -0,0 +1,39 @@
+# This workflow will install project dependencies and run formatter
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: formatter
+
+on:
+  push:
+    branches: ["develop"]
+  pull_request:
+    branches: ["*"]
+
+jobs:
+  format:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v3
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install graphviz
+        run: |
+          sudo apt-get install graphviz
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
+
+      - name: Format Python files with black
+        uses: psf/black@stable
```

Comment:
@caleb531 is this intended to check formatting on commit? If so, you should be using `flake8` with the `flake8-black` and `flake8-isort` plugins. Just running black without flake8 won't check for things like unused imports, for example.

Reply:
@eliotwrobson Probably a mistake due to my unfamiliarity with black. I've only ever used flake8+pycodestyle for the longest time in my projects, and thought that pycodestyle (a required dependency of flake8) would conflict with black in this case, hence the removal of flake8 and keep just pyflakes.

At any rate, since we haven't formatted the other files in the project yet, I reverted the addition of the CI action for now so I don't keep getting 'failed job' emails on every commit due to the formatter complaining about those files.

Reply:
Pycodestyle shouldn't conflict with black. It's a bit tricky, but I've actually set up something similar in another project. Basically, you need the following dependencies:
```
flake8-black
flake8-isort
Flake8-pyproject
flake8
black
isort
```
and then in `pyproject.toml` you need the following configurations
```
[tool.flake8]
# Black compatibility
max-line-length = 88
extend-ignore = ["E203", "W503"]

[tool.isort]
profile = "black"
```
This will essentially have `flake8` run `isort` and `black` on its own and check for inconsistencies with the way those tools do formatting. Then, you can have the workflows just run `flake8 .`. I would suggest having the formatting run before/after testing on the existing CI jobs, since this means not having to wait for setup to run twice for each Python version.

Reply:
>  I would suggest having the formatting run before/after testing on the existing CI jobs, since this means not having to wait for setup to run twice for each Python version.

I'd prefer to have a separate GitHub action because if the `tests` action fails, I want to assume without issue that a test failed/errored, rather than just a formatting issue or linter error creeping in.

> Like I said above, I think this is best done in a different PR, where we can run the autoformatting over everything as well.

What do you mean? Should I disable the format-on-file-save behavior so that the other files don't get auto-formatted? I thought you were going to run everything through black (i.e. format everything) anyway at the end of this PR. If that's the case, then what would be the purpose of the second PR?

Reply:
@eliotwrobson 

>  I would suggest having the formatting run before/after testing on the existing CI jobs, since this means not having to wait for setup to run twice for each Python version.

I'd prefer to have a separate GitHub action because if the `tests` action fails, I want to assume without issue that a test failed/errored, rather than just a formatting issue or linter error creeping in.

> Like I said above, I think this is best done in a different PR, where we can run the autoformatting over everything as well.

What do you mean? Should I disable the format-on-file-save behavior so that the other files don't get auto-formatted? I thought you were going to run everything through black (i.e. format everything) anyway at the end of this PR. If that's the case, then what would be the purpose of the second PR?

Reply:
@caleb531 

> What do you mean? Should I disable the format-on-file-save behavior so that the other files don't get auto-formatted? I thought you were going to run everything through black (i.e. format everything) anyway at the end of this PR. If that's the case, then what would be the purpose of the second PR?

Actually, on second thought I think it's fine. My thinking is that running the autoformatting right now is probably not the best idea since if something like #129 gets merged before this, it will be difficult to deal with merge conflicts because the base branch will have non-auto-formatted code. But if we just wait to run then it should be alright.

Reply:
@eliotwrobson I see, that makes sense. I actually already ran black on automaton.py in a recent commit (when I added the VS Code configuration to format on save), but I'm pushing a commit now to disable that until a later point.